### PR TITLE
Leafover of #894

### PR DESCRIPTION
**Pull request by HoshinoTented**
* Update the grammar and we can eliminate the ambiguities by `using` syntax.
* Update the grammar so that we can re-export a module by `public import`
* Open a module will import the modules that were exported.
* Fix diamond inheritance
* The symbols defined in `This` module will shadow the symbols imported from other modules, whatever the order is.
* Fix legacy code.

### CODE REVIEW

File: base/src/main/java/org/aya/resolve/visitor/StmtShallowResolver.java [hunk 1]
```diff
@@ -76,7 +76,8 @@ public void resolveStmt(@NotNull Stmt stmt, @NotNull ModuleContext context) {
           var symbol = ctx.modules().get(mod).symbols().getMaybe(use.id().component(), use.id().name());
           assert symbol.isOk();   // checked in openModule
           var defVar = symbol.get();
-          var renamedOpDecl = new ResolveInfo.RenamedOpDecl(new OpDecl.OpInfo(use.asName(), use.asAssoc()));
+          var asName = use.asName().getOrDefault(use.id().name());      // TODO: probably incorrect
+          var renamedOpDecl = new ResolveInfo.RenamedOpDecl(new OpDecl.OpInfo(asName, use.asAssoc()));
```
Comment (HoshinoTented) [inline base/src/main/java/org/aya/resolve/visitor/StmtShallowResolver.java:80]: I am not sure whether this is correct. I have no idea about `ResolveInfo` and `RenamedOpDecl`.

File: base/src/main/java/org/aya/resolve/context/ModuleSymbol.java [hunk 2]
```diff
@@ -86,10 +87,31 @@ public boolean contains(@NotNull String unqualified) {
...
+      case 0 -> Error.NotFound;
+      case 1 -> null;
+      default -> Error.Ambiguous;
+    };
+  }
+
+  /**
+   * @return null if contains
+   */
+  public @Nullable Error containsDefinitely(@NotNull ModulePath component, @NotNull String unqualified) {
```
Comment (ice1000) [inline base/src/main/java/org/aya/resolve/context/ModuleSymbol.java:105]: This is unused, right? Why

**Comment by ice1000** [comment]
image

😱

File: base/src/main/java/org/aya/resolve/context/ModuleSymbol.java (hunk 2 above)
Comment (HoshinoTented) [reply base/src/main/java/org/aya/resolve/context/ModuleSymbol.java:105]: I just think it should exist.

File: base/src/main/java/org/aya/resolve/visitor/StmtShallowResolver.java [hunk 3]
```diff
@@ -36,6 +37,13 @@ public void resolveStmt(@NotNull Stmt stmt, @NotNull ModuleContext context) {
     switch (stmt) {
       case Decl decl -> resolveDecl(decl, context);
       case Command.Module mod -> {
+        var wholeModeName = context.moduleName().appended(mod.name());
+        // Is there a module with path {context.moduleName}::{mod.name} ?
+        if (loader.exists(wholeModeName)) {
+          // 🥲
+          context.reportAndThrow(new NameProblem.DuplicateModNameError(ModulePath.qualified(wholeModeName), mod.sourcePos()));
```
Comment (imkiva) [inline base/src/main/java/org/aya/resolve/visitor/StmtShallowResolver.java:44]: It is good to have a new problem saying something like `The inner module ${wholeModName} clashes with some file level module ${wholeModName.replace("::", "/")}.aya`, which looks more friendly.

**Review by imkiva — APPROVED** [review]
Some questions

File: base/src/main/java/org/aya/resolve/context/ModuleExport.java [hunk 4]
```diff
@@ -66,68 +81,132 @@ public ModuleExport() {
...
+
+  public void putAll(@NotNull ModuleExport other) {
+    symbols.table().putAll(other.symbols.table());
+    modules.putAll(other.modules);
+  }
+
+  public void clear() {
+    symbols.table().clear();
+    modules.clear();
+  }
```
[the remark this answers is not in the document]
Comment (HoshinoTented) [reply base/src/main/java/org/aya/resolve/context/ModuleExport.java:175]: I think this is REPL-only. @ice1000

File: cli-impl/src/main/java/org/aya/cli/interactive/ReplContext.java [hunk 5]
```diff
@@ -33,7 +33,7 @@ public ReplContext(@NotNull Context parent, @NotNull ImmutableSeq<String> name)
   ) {
     // REPL always overwrites symbols.
     symbols().add(modName, name, ref);
-    // REPL exports nothing, because nothing can import REPL.
+    if (ref instanceof DefVar<?, ?> defVar && acc == Stmt.Accessibility.Public) exportSymbol(modName, name, defVar);
   }
```
[the remark this answers is not in the document]
Comment (HoshinoTented) [reply cli-impl/src/main/java/org/aya/cli/interactive/ReplContext.java:37]: A data/struct is also a ReplContext in REPL which exports its ctors/fields.

File: base/src/main/java/org/aya/resolve/context/ModuleExport.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (HoshinoTented) [reply base/src/main/java/org/aya/resolve/context/ModuleExport.java:175]: So does `ModuleExport#putAll`

**Comment by ice1000** [comment]
bors d=HoshinoTented

File: cli-impl/src/main/java/org/aya/cli/interactive/ReplContext.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (imkiva) [reply cli-impl/src/main/java/org/aya/cli/interactive/ReplContext.java:37]: Thank you!!

**Pull request merged**